core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1163,4 +1170,21 @@ public long getLastWriteNanos() {
      * This operation will close the engine if the recovery fails.
      */
     public abstract Engine recoverFromTranslog() throws IOException;
+
+    /**
+     * Called when a refresh includes the location.
+     */
+    public static interface RefreshListener {
+        /**
+         * The location to wait for. Must not return null and should be fast.
+         */
+        Translog.Location location();
```

Comment:
I think essentially we don't need any additional class no? `Location` already exits, `Runnable` is a JDK class. Internally we can use `Tuple` to hold the two so we have an additiional instance which is fine I guess. The semantics are really clear here if you have Runnable, it's called once it's visible. No methods that return anything or take any arguments - simple?